price-server/src/app/mod.rs

```diff
@@ -1,43 +1,137 @@
+            Self::subscribe_kollider(
+                pubsub_cfg.clone(),
+                health_check_trigger.clone(),
+                kollider_price_cache.clone(),
+            )
+            .await?;
+            providers.insert(
+                KOLLIDER_EXCHANGE_ID.to_string(),
+                (Box::new(kollider_price_cache), config.weight),
+            );
+        }
+
+        if let Some(config) = exchanges_cfg.okex.as_ref() {
+            let okex_price_cache = ExchangeTickCache::new(Duration::seconds(30));
+            Self::subscribe_okex(
+                pubsub_cfg.clone(),
+                health_check_trigger.clone(),
+                okex_price_cache.clone(),
+            )
+            .await?;
+            providers.insert(
+                OKEX_EXCHANGE_ID.to_string(),
+                (Box::new(okex_price_cache), config.weight),
+            );
+        }
+
+        let fee_calculator = FeeCalculator::new(fee_calc_cfg);
+        let app = Self {
+            price_mixer: PriceMixer::new(providers),
+            fee_calculator,
+        };
+
+        Ok(app)
+    }
+
+    async fn subscribe_okex(
+        pubsub_cfg: PubSubConfig,
+        health_check_trigger: Arc<RwLock<HealthCheckTrigger>>,
+        price_cache: ExchangeTickCache,
+    ) -> Result<(), PriceAppError> {
         let subscriber = Subscriber::new(pubsub_cfg).await?;
         let mut stream = subscriber.subscribe::<OkexBtcUsdSwapPricePayload>().await?;
         tokio::spawn(async move {
-            while let Some(check) = health_check_trigger.next().await {
+            while let Some(check) = health_check_trigger.write().await.next().await {
                 check
                     .send(subscriber.healthy(Duration::seconds(20)).await)
                     .expect("Couldn't send response");
             }
         });
 
-        let price_cache = ExchangePriceCache::new(Duration::seconds(30));
-        let fee_calculator = FeeCalculator::new(fee_calc_cfg);
-        let app = Self {
-            price_cache: price_cache.clone(),
-            fee_calculator,
-        };
+        let _ = tokio::spawn(async move {
+            while let Some(msg) = stream.next().await {
+                let span = info_span!(
+                    "price_tick_received",
+                    message_type = %msg.payload_type,
+                    correlation_id = %msg.meta.correlation_id
+                );
+                shared::tracing::inject_tracing_data(&span, &msg.meta.tracing_data);
+
+                async {
+                    price_cache
+                        .apply_update(msg.payload.0, msg.meta.correlation_id)
+                        .await;
+                }
+                .instrument(span)
+                .await;
+            }
+        });
+
+        Ok(())
+    }
+
+    async fn subscribe_kollider(
+        pubsub_cfg: PubSubConfig,
+        health_check_trigger: Arc<RwLock<HealthCheckTrigger>>,
```

Comment:
I don't think passing the `health_check_trigger` around behind a `RwLock` will work like this. This way _either_ this _or_ the other method will get a handle to the health checker... but we want both subscribers to be checked on each health_check... so you must spawn off the health check with clones of the subscribers or something like that.